### PR TITLE
Feature/43885 replace tslint with eslint

### PR DESCRIPTION
**Pull request by GuyHarwood**
tslint is deprecated in favor of [eslint for typescript.](https://github.com/typescript-eslint/typescript-eslint)

- [x] remove eslint and standard
- [x] add eslint, plugins and config
- [x] start with basic ruleset, fix errors and warnings
- [x] add jest linting plugin to eslint
- [x] introduce stricter ruleset via standardjs typescript plugin)
- [x] fix errors and warnings
- [x] introduce stricter policy for jest (recommended & style sets)
- [x] customise with additional policies
- [x] fix errors and warnings

Warning: there are a lot of changes in this PR, due to the new linting rules.  The key files for the lint setup are...
- package.json
- .eslintrc.js
- .eslintignore

### CODE REVIEW

File: tslib/package.json [hunk 1]
```diff
@@ -7,7 +7,8 @@
     "build": "tsc -b --v",
     "build:watch": "tsc -b --v -w",
     "rebuild": "yarn clean && yarn build",
-    "lint": "standard && tslint --project .",
+    "lint": "yarn eslint . --ext .js,.jsx,.ts,.tsx",
+    "lint:autofix": "yarn eslint . --ext .js,.jsx,.ts,.tsx --fix",
```
Comment (jon-shipley) [inline tslib/package.json:11]: Save some keystrokes and get closer parity to the original flag?

```suggestion
    "lint:fix": "yarn eslint . --ext .js,.jsx,.ts,.tsx --fix",
```

File: tslib/src/azure/storage-helper.ts [hunk 2]
```diff
@@ -97,15 +95,14 @@ export interface IBlobStorageHelper {
...
   constructor () {
     super(config.AzureStorage.ConnectionString)
   }
 
-  deleteContainerAsync (container: string): Promise<void> {
-    return new Promise((resolve, reject) => {
-      this.deleteContainer(container, (error, response) => {
-        if (error) {
+  async deleteContainerAsync (container: string): Promise<void> {
+    return await new Promise((resolve, reject) => {
```
Comment (jon-shipley) [inline tslib/src/azure/storage-helper.ts:103]: Shouldn't it be up to the caller to do the await'ing here? Awaiting on the return value of an async function is an anti-pattern.

File: tslib/src/functions/check-allocator/check-allocation.data.service.ts [hunk 3]
```diff
@@ -26,6 +26,7 @@ export class CheckAllocationDataService implements ICheckAllocationDataService {
   }
 
   async getFormsUsedByPupil (pupilUUID: string): Promise<any[]> {
+    console.log(pupilUUID)
```
Comment (jon-shipley) [inline tslib/src/functions/check-allocator/check-allocation.data.service.ts:29]: debug?

File: tslib/src/functions/check-allocator/check-form-allocation.service.ts [hunk 4]
```diff
@@ -5,6 +5,7 @@ export interface ICheckFormAllocationService {
 
 export class CheckFormAllocationService implements ICheckFormAllocationService {
   async allocate (pupilId: number): Promise<number> {
+    console.log(pupilId)
```
Comment (jon-shipley) [inline tslib/src/functions/check-allocator/check-form-allocation.service.ts:8]: debug?

**Review by jon-shipley — COMMENTED** [review]
Thanks for picking this work up! 🚀   I've reviewed 40 files, I'll review some more in another reading.

File: tslib/src/azure/storage-helper.ts (hunk 2 above)
Comment (GuyHarwood) [reply tslib/src/azure/storage-helper.ts:103]: that's what i thought, until i read [this](https://jakearchibald.com/2017/await-vs-return-vs-return-await/)

Comment (GuyHarwood) [reply tslib/src/azure/storage-helper.ts]: resolved - rules modified to disallow `return await` unless in `try..catch` block

File: tslib/src/azure/storage-helper.ts [hunk 5]
```diff
@@ -51,10 +49,10 @@ export class AsyncTableService extends az.TableService implements IAsyncTableSer
     })
   }
 
-  queryEntitiesAsync (table: string, tableQuery: az.TableQuery, currentToken: az.TableService.TableContinuationToken): Promise<Error | any> {
+  async queryEntitiesAsync (table: string, tableQuery: az.TableQuery, currentToken: az.TableService.TableContinuationToken): Promise<Error | any> {
```
Comment (jon-shipley) [inline tslib/src/azure/storage-helper.ts:52]: `async` is not required as it does not actually use an await.  Just returning a promise is enough for it to be await'd on.

**Review by jon-shipley — COMMENTED** [review]
LGSF. More reviewing.

File: tslib/src/azure/storage-helper.ts (hunk 5 above)
Comment (GuyHarwood) [reply tslib/src/azure/storage-helper.ts:52]: yes i'm aware of this, i usually omit the `async` keyword too.  This is the reasoning behind the rule - https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/promise-function-async.md

**Review by jon-shipley — COMMENTED** [review]
LGSF

**Comment by GuyHarwood** [comment]
@jon-shipley please could you pay particular attention to `tslib/src/functions/school-results-cache/services/result.service.ts` and the corresponding test suite, as this had to be heavily modified to satisfy the linter.

Most notably, an expectation wanted undefined, but we are returning null, as i updated the mapper.  Please check.

**Review by jon-shipley — COMMENTED** [review]
LGSF

File: tslib/src/functions/school-results-cache/services/result.service.spec.ts [hunk 6]
```diff
@@ -143,13 +191,61 @@ describe('result.service', () => {
...
 
     test('it sorts the pupils alphabetically - if the lastNames are the same it then sorts by foreName', async () => {
-      const mockResultData = [
-        { lastName: 'Smith', foreName: 'Toad' },
-        { lastName: 'Smith', foreName: 'Mario' },
-        { lastName: 'Anchovy', foreName: 'Zeus' }
+      const mockResultData: IRawPupilResult[] = [
+        {
+          lastName: 'Smith',
+          foreName: 'Toad',
```
Comment (jon-shipley) [inline tslib/src/functions/school-results-cache/services/result.service.spec.ts:197]:
```suggestion
          foreName: 'Thomas',
```

File: tslib/src/functions/school-results-cache/services/result.service.spec.ts [hunk 7]
```diff
@@ -143,13 +191,61 @@ describe('result.service', () => {
...
+          currentCheckId: 2,
+          mark: 1,
+          pupilId: 2,
+          school_id: 1,
+          urlSlug: 'urlSlug',
+          dateOfBirth: moment('2012-01-01'),
+          middleNames: ''
+        },
+        {
+          lastName: 'Anchovy',
```
Comment (jon-shipley) [inline tslib/src/functions/school-results-cache/services/result.service.spec.ts:230]:
```suggestion
          lastName: 'Adam',
```

**Review by jon-shipley — COMMENTED** [review]
Review of school-results-cache/services/results.service.ts and the spec.  Looks good to me, but I have made some suggested edits to help the code read better, for those who do not know their Windows SQL Clients.

File: tslib/src/functions/check-pin-expiry/check-pin-expiry.service.spec.ts [hunk 8]
```diff
@@ -23,7 +23,6 @@ describe('CheckPinExpiryService', () => {
 
   test('modifies the database', async () => {
     await sut.process()
-    expect(sqlServiceMock.modify).toHaveBeenCalled()
-    expect(sqlServiceMock.modify.mock.calls[0][0]).toMatch(/^DELETE FROM \[mtc_admin\]\.\[checkPin\]/)
+    expect(sqlServiceMock.modify).toHaveBeenCalledWith(expect.any(String), [])
```
Comment (jon-shipley) [inline tslib/src/functions/check-pin-expiry/check-pin-expiry.service.spec.ts:26]: I prefer the first test rather than the replacement.  Why the change?

**Review by jon-shipley — APPROVED** [review]
LGTM

File: tslib/src/functions/check-pin-expiry/check-pin-expiry.service.spec.ts (hunk 8 above)
Comment (GuyHarwood) [reply tslib/src/functions/check-pin-expiry/check-pin-expiry.service.spec.ts:26]: i changed that because it is verifying something that is not its concern (the underlying storage mechanism).  Ideally there should be a `checkPinExpiryDataService.deleteCheckPin` method which it could verify, but sorting that is out of scope.

However, the first/original test is more specific, and provides more value really, so i'm happy to change it back 👍

**Review by jon-shipley — APPROVED** [review]
